lib/linguist/heuristics.yml

```diff
@@ -61,6 +61,10 @@ disambiguations:
     - pattern: '\b(?i:(CODEUNIT|PAGE|PAGEEXTENSION|PAGECUSTOMIZATION|DOTNET|ENUM|ENUMEXTENSION|VALUE|QUERY|REPORT|TABLE|TABLEEXTENSION|XMLPORT|PROFILE|CONTROLADDIN|REPORTEXTENSION|INTERFACE|PERMISSIONSET|PERMISSIONSETEXTENSION|ENTITLEMENT))\b'
   # Open-ended fallback to Perl AutoLoader
   - language: Perl
+- extensions: ['.app']
+  rules:
+  - language: Erlang
+    pattern: '^{\s*(?:application|''application'')\s*,\s*(?:[a-z]+[a-z\d_@]*|''[\s\S]+'')\s*,\s*\[[\s\S]*\]\s*}\.$'
```

Comment:
1. The sequence `'[\s\S]+'` will match stuff like `'Foo's Bar'`, which I'm assuming isn't valid Erlang (I know nothing of the language). I'm assuming you meant `'.+'` where `.` includes newlines as well (i.e., the "dotall" modifier typical of most regex flavours). If so, I recommend using `'[^']+'` instead (which doesn't account for escape sequences, if such a thing is supported).

2. Did you intend for `[a-z]` to match only *lowercase* characters? If not, `[a-z\d_@]` can be simplified into `[\w@]`, as (IIRC) GitHub's Linguist process runs without full [CLDR](https://cldr.unicode.org/) support, meaning `\w` will match its traditional POSIX definition (`[[:alnum:]]`, or `[A-Za-z0-9_]`). @lildude will correct me if I'm wrong, I'm sure. 😉

3. `\.$` doesn't account for trailing whitespace. Unless that's a deliberate omission, I recommend using `\.[ \t]*$` to match EOL.

Putting the changes proposed by all three points together, we get:

```suggestion
-    pattern: '^{\s*(?:application|''application'')\s*,\s*(?:[a-z]+[a-z\d_@]*|''[\s\S]+'')\s*,\s*\[[\s\S]*\]\s*}\.$'
+    pattern: '(?m)^{\s*(?:application|''application'')\s*,\s*(?:[a-z]+[\w@]*|''[^'']+'')\s*,\s*\[.*?\]\s*}\.[\t]*$'
```

<table width="700"><tbody><tr><td valign="top" width="350"><details><summary><b>Source (readable)</b></summary>

~~~regexp
(?xm) ^
{ \s*
(?:
	application
	|
	'application'
)
\s* , \s*
(?:
	[a-z]+[\w@]*
	|
	'[^']+'
)
\s* , \s*
\[ .*? \]
\s* }
\. [ \t]* $
~~~

</details></td><td valign="top" width="350"><details><summary><b>Source (YAML string)</b></summary>

~~~yaml
'(?m)^{\s*(?:application|''application'')\s*,\s*(?:[a-z]+[\w@]*|''[^'']+'')\s*,\s*\[.*?\]\s*}\.[\t]*$'
~~~

</details></td></tr></tbody></table>

Reply:
1. Your assumption is correct.
2. The first `[a-z]` occurrence should only match lowercase letters. However, `[a-z\d_@]` can be reduced to `[\w@]` since, although uncommon, that particular segment of an atom _can_ include uppercase characters.
3. Not deliberate.

Thank you for your time. I have updated the code to reflect your suggestions.